File: scripts/generate_docx.py
"""
Docx Master — python-docx generator
Usage: python generate_docx.py <output.docx> <content.json>

Auto-installs python-docx if missing, then generates a .docx from JSON.
"""

import json
import subprocess
import sys
import os
from pathlib import Path

# ── Auto-install python-docx if missing ──────────────────────────
try:
    from docx import Document
    from docx.shared import Pt, Inches, Cm, Emu, RGBColor
    from docx.enum.text import WD_ALIGN_PARAGRAPH
    from docx.enum.style import WD_STYLE_TYPE
    from docx.oxml.ns import qn
except ImportError:
    print("[docx-master] python-docx not found, installing...")
    subprocess.check_call(
        [sys.executable, "-m", "pip", "install", "python-docx", "-q"],
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
    )
    print("[docx-master] python-docx installed OK")
    from docx import Document
    from docx.shared import Pt, Inches, Cm, Emu, RGBColor
    from docx.enum.text import WD_ALIGN_PARAGRAPH
    from docx.enum.style import WD_STYLE_TYPE
    from docx.oxml.ns import qn

# ── Configuration ─────────────────────────────────────────────────
FONT_EAST_ASIA = "宋体"
FONT_ASCII = "Times New Roman"

ALIGN_MAP = {
    "left":   WD_ALIGN_PARAGRAPH.LEFT,
    "center": WD_ALIGN_PARAGRAPH.CENTER,
    "right":  WD_ALIGN_PARAGRAPH.RIGHT,
}


def setup_styles(doc):
    """Chinese-friendly defaults: 宋体 body, sized headings 1-4."""
    style = doc.styles["Normal"]
    style.font.name = FONT_ASCII
    style.font.size = Pt(12)
    style.element.rPr.rFonts.set(qn("w:eastAsia"), FONT_EAST_ASIA)
    style.paragraph_format.space_after = Pt(6)
    style.paragraph_format.line_spacing = 1.35

    for i, sz in {1: 18, 2: 14, 3: 13, 4: 12}.items():
        hname = f"Heading {i}"
        if hname in [s.name for s in doc.styles if s.type == WD_STYLE_TYPE.PARAGRAPH]:
            hs = doc.styles[hname]
        else:
            hs = doc.styles.add_style(hname, WD_STYLE_TYPE.PARAGRAPH)
            hs.base_style = doc.styles["Normal"]
        hs.font.name = FONT_ASCII
        hs.element.rPr.rFonts.set(qn("w:eastAsia"), FONT_EAST_ASIA)
        hs.font.bold = True
        hs.font.size = Pt(sz)


def _font_run(run):
    run.font.name = FONT_ASCII
    run._element.rPr.rFonts.set(qn("w:eastAsia"), FONT_EAST_ASIA)


# ── Content builders ──────────────────────────────────────────────

def add_paragraph(doc, text, bold=False, italic=False, align=None,
                  sz=None, color=None):
    p = doc.add_paragraph()
    run = p.add_run(text)
    run.bold = bold
    run.italic = italic
    _font_run(run)
    if sz:
        run.font.size = Pt(int(sz) / 2)   # half-pts → pt
    if color:
        run.font.color.rgb = RGBColor.from_string(color)
    if align in ALIGN_MAP:
        p.alignment = ALIGN_MAP[align]


def add_bullet(doc, text):
    p = doc.add_paragraph(style="List Bullet")
    p.clear()
    run = p.add_run(text)
    _font_run(run)


def add_heading(doc, text, level=1):
    h = doc.add_heading(text, level=level)
    for run in h.runs:
        _font_run(run)


def add_table(doc, headers, rows, col_widths=None):
    table = doc.add_table(rows=1 + len(rows), cols=len(headers))
    table.style = "Table Grid"

    for i, txt in enumerate(headers):
        cell = table.rows[0].cells[i]
        cell.text = ""
        run = cell.paragraphs[0].add_run(txt)
        run.bold = True
        _font_run(run)
        run.font.size = Pt(11)
        cell.paragraphs[0].alignment = WD_ALIGN_PARAGRAPH.CENTER

    for r, row_data in enumerate(rows):
        for c, txt in enumerate(row_data):
            cell = table.rows[r + 1].cells[c]
            cell.text = ""
            run = cell.paragraphs[0].add_run(str(txt))
            _font_run(run)
            run.font.size = Pt(10.5)

    if col_widths:
        for row in table.rows:
            for i, w in enumerate(col_widths):
                row.cells[i].width = Emu(w)

    doc.add_paragraph()   # spacer


def add_image(doc, image_path, width=None, height=None, caption=None):
    if not os.path.isfile(image_path):
        print(f"[docx-master] WARNING: image not found: {image_path}")
        return

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run()

    kwargs = {}
    if width:
        kwargs["width"] = Emu(width)
    if height:
        kwargs["height"] = Emu(height)
    if not kwargs:
        kwargs["width"] = Inches(5.5)

    try:
        run.add_picture(image_path, **kwargs)
    except Exception as exc:
        print(f"[docx-master] WARNING: cannot embed {image_path}: {exc}")
        return

    if caption:
        cp = doc.add_paragraph()
        cp.alignment = WD_ALIGN_PARAGRAPH.CENTER
        cr = cp.add_run(caption)
        cr.font.size = Pt(10)
        cr.font.color.rgb = RGBColor(0x66, 0x66, 0x66)
        _font_run(cr)


# ── Main ──────────────────────────────────────────────────────────

CONTENT_HANDLERS = {
    "paragraph": lambda doc, item: add_paragraph(
        doc, item["text"],
        bold=item.get("bold", False),
        italic=item.get("italic", False),
        align=item.get("align"),
        sz=item.get("sz"),
        color=item.get("color"),
    ),
    "heading": lambda doc, item: add_heading(
        doc, item["text"], item.get("level", 2),
    ),
    "bullet": lambda doc, item: add_bullet(doc, item["text"]),
    "table": lambda doc, item: add_table(
        doc, item["headers"], item["rows"], item.get("colWidths"),
    ),
    "image": lambda doc, item: add_image(
        doc, item["path"],
        item.get("width"), item.get("height"), item.get("caption"),
    ),
}


def main():
    if len(sys.argv) < 3:
        print("Usage: python generate_docx.py <output.docx> <content.json>")
        sys.exit(1)

    out_path = sys.argv[1]
    with open(sys.argv[2], "r", encoding="utf-8") as fh:
        content = json.load(fh)

    doc = Document()
    setup_styles(doc)

    # Document title
    if content.get("title"):
        t = doc.add_heading(content["title"], level=0)
        t.alignment = WD_ALIGN_PARAGRAPH.CENTER
        for run in t.runs:
            _font_run(run)

    # Sections
    img_count = 0
    for sec in content.get("sections", []):
        if sec.get("heading"):
            add_heading(doc, sec["heading"], sec.get("level", 1))
        for item in sec.get("content", []):
            handler = CONTENT_HANDLERS.get(item.get("type"))
            if handler:
                handler(doc, item)
                if item.get("type") == "image":
                    img_count += 1

    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
    doc.save(out_path)
    print(f"OK: {out_path} ({os.path.getsize(out_path)} bytes, {img_count} images)")


if __name__ == "__main__":
    main()

File: scripts/generate_docx.js
/**
 * Docx Master — Node.js + JSZip fallback generator
 * Usage: node generate_docx.js <output.docx> <content.json>
 *
 * Auto-installs JSZip if missing, then generates a .docx from JSON.
 * Works on any platform with Node.js ≥ 12.
 */

const fs = require('fs');
const path = require('path');
const cp = require('child_process');

// ── Auto-install JSZip if missing ────────────────────────────────
let JSZip;
try {
  JSZip = require('jszip');
} catch (e) {
  console.log('[docx-master] jszip not found, installing...');
  try {
    cp.execSync('npm install jszip --no-save', {
      cwd: __dirname,
      stdio: 'pipe',
      timeout: 30000,
    });
    JSZip = require('jszip');
    console.log('[docx-master] jszip installed OK');
  } catch (e2) {
    console.error('[docx-master] ERROR: cannot install jszip.');
    console.error('  Install manually: npm install jszip');
    process.exit(1);
  }
}

// ── CLI ───────────────────────────────────────────────────────────
const args = process.argv.slice(2);
if (args.length < 2) {
  console.error('Usage: node generate_docx.js <output.docx> <content.json>');
  process.exit(1);
}

const outPath = args[0];
const content = JSON.parse(fs.readFileSync(args[1], 'utf-8'));

// ── XML helpers ───────────────────────────────────────────────────

function esc(s) {
  return String(s)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function pNode(text, opts = {}) {
  const bold = opts.bold ? '<w:rPr><w:b/></w:rPr>' : '';
  const sz = opts.sz ? `<w:rPr><w:sz w:val="${opts.sz}"/></w:rPr>` : '';
  const italic = opts.italic ? '<w:rPr><w:i/></w:rPr>' : '';
  const color = opts.color ? `<w:rPr><w:color w:val="${opts.color}"/></w:rPr>` : '';
  const align = opts.align ? `<w:pPr><w:jc w:val="${opts.align}"/></w:pPr>` : '';
  const indent = opts.indent ? `<w:pPr><w:ind w:left="${opts.indent}" w:hanging="360"/></w:pPr>` : '';
  const rPr = bold || sz || italic || color;
  return `<w:p>${align}${indent}<w:r>${rPr}<w:t xml:space="preserve">${esc(text)}</w:t></w:r></w:p>`;
}

function headingNode(text, level) {
  const sizes = { 1: 36, 2: 28, 3: 24, 4: 22 };
  const styleIds = { 1: '2', 2: '3', 3: '4', 4: '5' };
  const sz = sizes[level] || 22;
  const sid = styleIds[level] || '5';
  return `<w:p><w:pPr><w:pStyle w:val="${sid}"/></w:pPr><w:r><w:rPr><w:b/><w:sz w:val="${sz}"/></w:rPr><w:t>${esc(text)}</w:t></w:r></w:p>`;
}

function drawingXml(id, rid, cx, cy) {
  return `<w:r><w:drawing><wp:inline distT="0" distB="0" distL="114300" distR="114300">
    <wp:extent cx="${cx}" cy="${cy}"/><wp:docPr id="${id}" name="img${id}"/>
    <a:graphic xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main">
      <a:graphicData uri="http://schemas.openxmlformats.org/drawingml/2006/picture">
        <pic:pic xmlns:pic="http://schemas.openxmlformats.org/drawingml/2006/picture">
          <pic:nvPicPr><pic:cNvPr id="${id}" name="img${id}"/><pic:cNvPicPr><a:picLocks noChangeAspect="1"/></pic:cNvPicPr></pic:nvPicPr>
          <pic:blipFill><a:blip r:embed="${rid}"/><a:stretch><a:fillRect/></a:stretch></pic:blipFill>
          <pic:spPr><a:xfrm><a:off x="0" y="0"/><a:ext cx="${cx}" cy="${cy}"/></a:xfrm><a:prstGeom prst="rect"><a:avLst/></a:prstGeom></pic:spPr>
        </pic:pic>
      </a:graphicData>
    </a:graphic>
  </wp:inline></w:drawing></w:r>`;
}

// ── Table helpers ─────────────────────────────────────────────────

function tableXml(headers, rows, colWidths) {
  const nCols = headers.length;
  const widths = colWidths || headers.map(() => Math.floor(8304 / nCols));
  let grid = '<w:tblGrid>';
  widths.forEach(w => { grid += `<w:gridCol w:w="${w}"/>`; });
  grid += '</w:tblGrid>';

  let h = `<w:tbl><w:tblPr><w:tblW w:w="0" w:type="auto"/><w:tblLayout w:type="fixed"/></w:tblPr>${grid}`;
  h += trXml(headers, true);
  rows.forEach(r => { h += trXml(r, false); });
  h += '</w:tbl>';
  return h;
}

function trXml(cells, isHeader) {
  let r = '<w:tr>';
  cells.forEach(c => {
    r += '<w:tc><w:tcPr><w:vAlign w:val="center"/></w:tcPr>';
    r += '<w:p><w:pPr><w:jc w:val="center"/></w:pPr>';
    r += isHeader
      ? `<w:r><w:rPr><w:b/></w:rPr><w:t>${esc(c)}</w:t></w:r>`
      : `<w:r><w:t>${esc(c)}</w:t></w:r>`;
    r += '</w:p></w:tc>';
  });
  r += '</w:tr>';
  return r;
}

// ── Build document body ───────────────────────────────────────────

let docBody = '';

if (content.title) {
  docBody += `<w:p><w:pPr><w:pStyle w:val="2"/><w:jc w:val="center"/></w:pPr><w:r><w:rPr><w:b/><w:sz w:val="36"/></w:rPr><w:t>${esc(content.title)}</w:t></w:r></w:p>`;
}

let imageIdx = 0;
const images = [];   // { path, rid }

(content.sections || []).forEach(sec => {
  if (sec.heading) {
    docBody += headingNode(sec.heading, sec.level || 1);
  }
  (sec.content || []).forEach(item => {
    switch (item.type) {
      case 'paragraph':
        docBody += pNode(item.text, item);
        break;
      case 'heading':
        docBody += headingNode(item.text, item.level || 2);
        break;
      case 'bullet':
        docBody += pNode(`• ${item.text}`, { indent: '720' });
        break;
      case 'table':
        docBody += tableXml(item.headers, item.rows, item.colWidths);
        docBody += pNode('', {});
        break;
      case 'image':
        imageIdx++;
        const rid = `rIdImg${imageIdx}`;
        const cx = item.width || 5400000;
        const cy = item.height || 3000000;
        docBody += `<w:p><w:pPr><w:jc w:val="center"/></w:pPr>${drawingXml(100 + imageIdx, rid, cx, cy)}</w:p>`;
        if (item.caption) {
          docBody += pNode(item.caption, { align: 'center', sz: '20', color: '666666' });
        }
        images.push({ path: item.path, rid });
        break;
    }
  });
});

// ── Supporting XML parts ──────────────────────────────────────────

const RELS_ROOT = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="word/document.xml"/>
</Relationships>`;

const CONTENT_TYPES = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>
  <Default Extension="xml" ContentType="application/xml"/>
  <Default Extension="png" ContentType="image/png"/>
  <Default Extension="jpg" ContentType="image/jpeg"/>
  <Default Extension="jpeg" ContentType="image/jpeg"/>
  <Override PartName="/word/document.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml"/>
  <Override PartName="/word/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.styles+xml"/>
</Types>`;

const STYLES = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:styles xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main">
  <w:style w:type="paragraph" w:styleId="2"><w:name w:val="Heading 1"/><w:pPr><w:spacing w:before="240" w:after="120"/></w:pPr><w:rPr><w:rFonts w:ascii="宋体" w:hAnsi="宋体"/><w:b/><w:sz w:val="36"/></w:rPr></w:style>
  <w:style w:type="paragraph" w:styleId="3"><w:name w:val="Heading 2"/><w:pPr><w:spacing w:before="200" w:after="100"/></w:pPr><w:rPr><w:rFonts w:ascii="宋体" w:hAnsi="宋体"/><w:b/><w:sz w:val="28"/></w:rPr></w:style>
  <w:style w:type="paragraph" w:styleId="4"><w:name w:val="Heading 3"/><w:pPr><w:spacing w:before="160" w:after="80"/></w:pPr><w:rPr><w:rFonts w:ascii="宋体" w:hAnsi="宋体"/><w:b/><w:sz w:val="24"/></w:rPr></w:style>
  <w:style w:type="paragraph" w:styleId="5"><w:name w:val="Heading 4"/><w:pPr><w:spacing w:before="120" w:after="60"/></w:pPr><w:rPr><w:rFonts w:ascii="宋体" w:hAnsi="宋体"/><w:b/></w:rPr></w:style>
</w:styles>`;

let docRels = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">`;
images.forEach(img => {
  docRels += `\n  <Relationship Id="${img.rid}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image" Target="media/image${img.rid.replace('rIdImg', '')}.png"/>`;
});
docRels += '\n</Relationships>';

// ── Package & output ──────────────────────────────────────────────

async function main() {
  const zip = new JSZip();

  // mimetype MUST be first, uncompressed
  zip.file('mimetype', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document', { compression: 'STORE' });

  zip.file('_rels/.rels', RELS_ROOT);
  zip.file('[Content_Types].xml', CONTENT_TYPES);
  zip.file('word/document.xml', `<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" xmlns:pic="http://schemas.openxmlformats.org/drawingml/2006/picture"><w:body>${docBody}</w:body></w:document>`);
  zip.file('word/styles.xml', STYLES);
  zip.file('word/_rels/document.xml.rels', docRels);

  // Embed images
  let imgNum = 0;
  for (const img of images) {
    imgNum++;
    try {
      const buf = fs.readFileSync(img.path);
      const ext = path.extname(img.path).slice(1) || 'png';
      zip.file(`word/media/image${imgNum}.${ext}`, buf);
    } catch (e) {
      console.error(`[docx-master] WARNING: cannot read image: ${img.path}`);
    }
  }

  const buf = await zip.generateAsync({ type: 'nodebuffer', compression: 'DEFLATE', compressionOptions: { level: 6 } });
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, buf);
  console.log(`OK: ${outPath} (${buf.length} bytes, ${images.length} images)`);
}

main().catch(e => { console.error(e); process.exit(1); });
